From ab939cc6862dcc3dca9cdae1b70ac94dbea7c20e Mon Sep 17 00:00:00 2001
From: Sivashankari Madhavan <quic_sivamadh@quicinc.com>
Date: Sun, 24 Apr 2022 18:55:29 +0530
Subject: [PATCH] Adds the iw command for green AP

Adds the iw command for Green AP support. Use the command,
iw wlanX set ap_ps 1 - For enable the green AP
iw wlanX set ap_ps 0 - For disable the green AP

Signed-off-by: Sivashankari Madhavan <quic_sivamadh@quicinc.com>
---
 interface.c | 32 ++++++++++++++++++++++++++++++++
 nl80211.h   |  5 ++++-
 2 files changed, 36 insertions(+), 1 deletion(-)

--- a/interface.c
+++ b/interface.c
@@ -1222,3 +1222,52 @@ COMMAND(set, tidconf, "[peer <MAC addres
 	"  $ iw dev wlan0 set tidconf peer xx:xx:xx:xx:xx:xx tids 0x2 bitrates auto\n"
 	"  $ iw dev wlan0 set tidconf peer xx:xx:xx:xx:xx:xx tids 0x2 bitrates limit legacy-5 ht-mcs-5 vht-mcs-5 he-mcs-5 2:3\n"
 	);
+
+static int handle_interface_ap_ps (struct nl80211_state *state,
+                                   struct nl_msg *msg,
+                                   int argc, char **argv,
+                                   enum id_input id)
+{
+	char *end;
+	uint8_t ap_ps = 0;
+	int i = 0;
+
+	if (argc == 3) {
+		if (!strcmp(argv[0], "-l")) {
+			unsigned int link_id;
+
+			link_id = strtol(argv[1], &end, 10);
+			if (*end)
+				return 1;
+			if (link_id < MAX_MLD_LINK) {
+				NLA_PUT_U8(msg, NL80211_ATTR_MLO_LINK_ID,
+						link_id);
+				i = 2;
+			}
+		} else {
+			printf("Invalid parameter: %s\n", argv[0]);
+			return 1;
+		}
+	} else {
+		if (argc != 1)
+			return 1;
+	}
+
+	ap_ps = strtol(argv[i], &end, 10);
+
+	if (*end)
+		return 2;
+
+	if (ap_ps > 1)
+		return -EINVAL;
+
+	NLA_PUT_U8(msg, NL80211_ATTR_AP_PS, ap_ps);
+
+	return 0;
+
+nla_put_failure:
+	return -ENOBUFS;
+}
+
+COMMAND(set, ap_ps, "[-l] <link id> <Enable[1]/Disable[0]>",
+                NL80211_CMD_UPDATE_AP, 0, CIB_NETDEV, handle_interface_ap_ps, NULL);
--- a/nl80211.h
+++ b/nl80211.h
@@ -1260,7 +1260,8 @@ enum nl80211_commands {
 	NL80211_CMD_DEL_KEY,
 
 	NL80211_CMD_GET_BEACON,
-	NL80211_CMD_SET_BEACON,
+	NL80211_CMD_UPDATE_AP,
+        NL80211_CMD_SET_BEACON = NL80211_CMD_UPDATE_AP,
 	NL80211_CMD_START_AP,
 	NL80211_CMD_NEW_BEACON = NL80211_CMD_START_AP,
 	NL80211_CMD_STOP_AP,
@@ -3225,6 +3226,8 @@ enum nl80211_attrs {
 
 	NL80211_ATTR_RU_PUNCT_BITMAP,
 
+	NL80211_ATTR_AP_PS,
+
 	/* add attributes here, update the policy in nl80211.c */
 
 	__NL80211_ATTR_AFTER_LAST,
--- a/iw.h
+++ b/iw.h
@@ -41,6 +41,7 @@ enum nlmsgerr_attrs {
 
 #define ETH_ALEN 6
 #define VHT_MUMIMO_GROUP_LEN 24
+#define MAX_MLD_LINK 15
 
 /* libnl 1.x compatibility code */
 #if !defined(CONFIG_LIBNL20) && !defined(CONFIG_LIBNL30)
